### ISSUE TITLE
Ability to conditionally set levels for lints when not using clippy through cargo features

### ISSUE THREAD

**Issue by crumblingstatue** [Manishearth/rust-clippy#574]
Certain projects might want to allow/warn/deny specific clippy lints. I, for example, want to allow `cyclomatic_complexity` for one of my projects, because it triggers a lot.

If I do `#![allow(cyclomatic_complexity)]`, rustc will warn about unknown lint when not compiling with clippy. I want to avoid this warning.

If one uses cargo features to enable clippy, one can conditionally set lint levels using `cfg_attr`, and the feature.

But there are other ways to run clippy, for example, using `cargo clippy`.

I am aware that one can supply command line arguments to for setting lint levels (e.g. `cargo clippy -- -A cyclomatic_complexity`), but I don't want to have to manually provide the arguments each time, and there could be multiple people working on the project, and I want everyone to have the same lint levels.

It would be nice to have a standard cfg variant for when the code is compiled with clippy.

If this is not desirable for some reason, what are some other alternatives?

I thought about instead `cargo clippy` setting a cfg variant, or maybe a configuration file for `cargo clippy` for setting lint levels for a project, but I'm not sure what's the best way to go about this.

**Comment by Manishearth** [comment]
The recommended way of handling this is cfg_attr, like you mentioned.

I thought of ways of doing this (I needed it myself for Servo) and there's very little that can be done short of a cargo feature per lint (this is hard since whilst you can codegen this with macros, there's no way to fill up the Cargo.toml with one and you need more python scripts). And then the Allow/Warn/Deny/Forbid choice mucks things up further.

> It would be nice to have a standard cfg variant for when the code is compiled with clippy.

I'm not quite sure what you mean by this?


A config file is a good idea, however.

Another option is to allow mutating lint levels from the plugin registrar. Then you just write a tiny plugin crate (this is easy) which calls into clippy's registrar and then modifies the lint levels it needs. We already do this in Servo anyway, minus the lint level mutation since it's not yet possible.

**Comment by crumblingstatue** [comment]
> > It would be nice to have a standard cfg variant for when the code is compiled with clippy.

> I'm not quite sure what you mean by this?

Isn't it possible for a compiler plugin (like clippy) to register a cfg value, so when code is compiled with that plugin active, that cfg variant is set?

For example, clippy would register `clippy`, so people could do `cfg_attr(clippy, allow(cyclomatic_complexity))`.

**Comment by Manishearth** [comment]
Nope.

**Comment by crumblingstatue** [comment]
In that case, I guess config files for `cargo clippy` are the way to go.

I'll open an issue on https://github.com/arcnmx/cargo-clippy, but I'll leave this one open for a bit in case people have more to say on this matter.

**Comment by Luthaf** [comment]
How about a `clippy.toml` file in the crate root? Is it possible for a compiler plugin to change the level of lints at startup from a configuration file?

Something like that:
```toml
clippy = "warn"
cyclomatic_complexity = "allow"
absurd_unsigned_comparisons = "deny"
```

**Comment by Manishearth** [comment]
> Is it possible for a compiler plugin to change the level of lints at startup from a configuration file?

No. It can be done at compile time or with a build script or something, feel free to try!

(Make sure the scripts in util/ continue to work with your new lint level system)

**Comment by mcarton** [comment]
The file should really be `Clippy.toml` with a capital letter à la `Cargo.toml` and `Makefile` :smile:

And while the file could not set lint levels, it could be used to configure some lints:
  * `cyclomatic_complexity`: what level is considered bad? ;
  * complex type: at what point is a type considered complex? ;
  * deactivate macro check: eg. in clippy we have a `if_let_chain` macro, some lints will never warn anything in that macro, it would be nice to be able to turn that off temporarily for review.

**Comment by Luthaf** [comment]
I am working on that, the design idea I have is to use a small compiler plugin [I wrote](https://github.com/Luthaf/confy) to read the configuration file at compile-time, and extract string values:
```rust
clippy_lint! {
    pub APPROX_CONSTANT,
    config!("approx_constant", "warn"),
    "the approximate of a known float constant (in `std::f64::consts` or `std::f32::consts`) \
     is found; suggests to use the constant"
}
```

Where the `clippy_lint!` macro replacing the `declare_lint!` macro
```rust
macro_rules! clippy_lint {
    (pub $name:ident, $level:expr, $desc:expr) => (
        pub static $name: &'static Lint = &Lint {
            name: stringify!($name),
            default_level: Level::from_str($level).unwrap(),
            desc: $desc,
        };
    );
}
```

Does it seems like a sound design to you?

My current issue is with the fact that all the lints must be static items, which does not make it easy to change them. I might end using some trick to generate the lint level at compile time, maybe with another compiler plugin. Do you have any idea of another way to do it?

**Comment by Manishearth** [comment]
I'd rather tweak the compiler so that the lint levels are mutable, this isn't a hard change. On the other hand, reading a config file at compile time has a host of problems.

**Comment by Luthaf** [comment]
Why not, if you think theses changes can be accepted. Which kind of problems do you think there can be?

**Comment by Manishearth** [comment]
You lose the ability to do [this](https://github.com/Manishearth/rust-clippy/pull/698/files#diff-04c6e90faac2675aa89e2176d2eec7d8R242).
